corgidrp/spec.py

```diff
@@ -0,0 +1,648 @@
+    (x_rot, y_rot) = rotate_points((xpts, ypts), -np.deg2rad(clock_ang), 
+                                    pivot_point = (xpts[refidx], ypts[refidx]))
+    
+    # Fit an intermediate polynomial to wavelength versus position
+    delta_x = (x_rot - x_rot[refidx]) * pixel_pitch_mm
+    pos_err = cent_errs * pixel_pitch_mm
+    weights = 1 / pos_err
+    lambda_func_x = np.poly1d(np.polyfit(x = delta_x, y = wavlens, deg = 2, w = weights))
+    # Determine the position at the reference wavelength
+    poly_roots = (np.poly1d(lambda_func_x) - ref_wavlen).roots
+    real_roots = poly_roots[np.isreal(poly_roots)]
+    root_select_ind = np.argmin(np.abs(poly_roots[np.isreal(poly_roots)]))
+    pos_ref = np.real(real_roots[root_select_ind])
+    np.testing.assert_almost_equal(lambda_func_x(pos_ref), ref_wavlen)
+    displacements_mm = delta_x - pos_ref
+
+    # Fit two polynomials:  
+    # 1. Displacement from the band center along the dispersion axis as a 
+    #    function of wavelength  
+    # 2. Wavelength as a function of displacement along the dispersion axis
+    (pfit_pos_vs_wavlen,
+     cov_pos_vs_wavlen) = np.polyfit(x = (wavlens - ref_wavlen) / ref_wavlen,
+                                      y = displacements_mm, deg = 3, w = weights, cov=True)
+    
+    (pfit_wavlen_vs_pos,
+     cov_wavlen_vs_pos) = np.polyfit(x = displacements_mm, y = wavlens, deg = 3, 
+                                      w = weights, cov=True)
+
+    return pfit_pos_vs_wavlen, cov_pos_vs_wavlen, pfit_wavlen_vs_pos, cov_wavlen_vs_pos
+
+
+def calibrate_dispersion_model(centroid_psf, band_center_file, prism = 'PRISM3', pixel_pitch_um = 13.0):
+    """ 
+    Generate a DispersionModel of the spectral dispersion profile of the CGI ZOD prism.
+
+    Args:
+       centroid_psf (SpectroscopyCentroidPsf): instance of SpectroscopyCentroidPsf calibration class
+       band_center_file (str): file name of the band centers
+       prism (str): Label for the selected DPAM zero-deviation prism; must be
+                    either 'PRISM3' or 'PRISM2'
+       pixel_pitch_um (float): EXCAM pixel pitch in micron, default 13 micron
+    
+    Returns:
+       data.DispersionModel: DispersionModel calfile object with the fit results including errors of the spectral trace and the dispersion
+    """
+    #if not isinstance(dataset, Dataset):
+    #    raise TypeError("Input must be a corgidrp.data.Dataset object.")
+    if prism not in ['PRISM2', 'PRISM3']:
+        raise ValueError("prism must be PRISM2 or PRISM3")
+    
+    if prism == 'PRISM2':
+        subband_list = ['2A', '2B', '2C']
+        ref_wavlen = 660.0
+        ref_cfam = '2'
+        #bandpass = [610, 710]
+        #halfheight = 10
+    else:
+        subband_list = ['3A', '3B', '3C', '3D', '3E', '3G']
+        ref_wavlen = 730.0
+        ref_cfam = '3'
+        #bandpass = [675, 785]
+        #halfheight = 30
+        
+    filters = centroid_psf.pri_hdr['FILTERS'].upper().split(",")
```

Comment:
I couldn't figure out who writes "FILTERS" into the primary header. this isn't a keyword by default.  Perhaps we need to construct a list of filters from the CFAMNAME keyword in from the L1 headers. @neilzim may have added this keyword for convenience into the test files, but we may not get it for the real data

Reply:
I set this "FILTERS" list in the test filter sweep image cube for convenience. It would be preferable to have the test operate on a list of individual image files in the expected format.